auth/DERMethods.m

```diff
@@ -0,0 +1,106 @@
+classdef DERMethods
+    % Class DERMethods contains all methods pertaining to
+    % Distinguished Encoding Rules (DER) and converting them to naked
+    % formats without the DER headers.
+
+    properties (Constant, Access = protected)
+        DER_ECDSA256_PK_HEADER = uint8([48; 89; 48; 19; 6; 7; 42; 134; 72; 206; 61; 2; 1; 6; 8; 42; 134; 72; 206; ...
+                                        61; 3; 1; 7; 3; 66; 0; 4])
+    end
+
+    methods (Static, Access = public)
+
+        function public_key = DER2PK(der_key, key_type)
+            % strips DER key of all headers and outputs naked public
+            % key  [r; s]
+            switch key_type
+                case "ECDSA256"
+                    if ~all(der_key(1:27) == DERMethods.DER_ECDSA256_PK_HEADER)
+                        error('DERMethods:DER2PK:BadHeader', 'Input header was not expected.');
+                    end
+                    public_key = der_key(28:end);
+                otherwise
+                    error('DERMethods:DER2PK:BadKeyType', 'Key type %s not implemented.', key_type);
+            end
+        end
+
+        function der_key = PK2DER(public_key, key_type)
+            % assembles DER signature from naked public key
+            switch key_type
+                case "ECDSA256"
+                    if length(public_key) ~= 64
+                        error('DERMethods:PK2DER:BadPublicKey', 'Input public key invalid.');
+                    end
+                    der_key = [DERMethods.DER_ECDSA256_PK_HEADER; public_key];
+                otherwise
+                    error('DERMethods:PK2DER:BadKeyType', 'Key type %s not implemented.', key_type);
+            end
+        end
+
+        function public_signature = DER2SIG(der_signature, key_type)
+            % strips DER signature of all headers and outputs naked public
+            % key signature [r; s]
+
+            switch key_type
+                case "ECDSA256"
```

Comment:
could it be more efficient to have a generic check der method since it seems like the exceptions code looks repetitive throughout the file

Reply:
certainly, the error ID codes are slightly different based on the function name...so if there was any easy way to call the function name, then one function could handle all these errors.

Efficient probably isn't the word you mean. I think you mean "brittle". Issues would arise when a programmer decides to change one error but forgets to change the others. Having the same function handle all the errors falls under a defensive coding principle.